File: Sources/App/Models/Item.swift
//
//  Item.swift
//  FinchServer
//
//  Created by Stefan Klein Nulent on 26/01/2025.
//

import Vapor
import Fluent

final class Item: Model, @unchecked Sendable, Comparable {
    
    // MARK: - Constants
    
    static let schema = "items"
    
    
    // MARK: - Private Properties
    
    private var orderIndex: Int {
        let disc = disc ?? 1
        return disc * 1000 + (track ?? 0)
    }
    
    
    
    // MARK: - Properties
    
    @ID(custom: .id, generatedBy: .database)
    var id: Int?
    
    @Field(key: "track")
    var track: Int?
    
    @Field(key: "disc")
    var disc: Int?
    
    @Field(key: "title")
    var title: String
    
    @Field(key: "artist")
    var artist: String
    
    @Field(key: "artists")
    var artists: String
    
    @Field(key: "path")
    var pathData: Data
    
    @Field(key: "length")
    var length: Double
    
    @Field(key: "format")
    var format: String
    
    @Field(key: "bitrate")
    var bitrate: Int
    
    @Field(key: "samplerate")
    var sampleRate: Int
    
    @Field(key: "album_id")
    var albumId: Int?
    
    @Field(key: "albumtype")
    var albumType: String?
    
    @OptionalParent(key: "album_id")
    var album: Album?
    
    @Field(key: "genre")
    var genre: String?
    
    @Field(key: "lyricist")
    var lyricist: String?
    
    @Field(key: "composer")
    var composer: String?
    
    @Field(key: "comments")
    var comments: String?
    
    @Field(key: "mb_trackid")
    var musicBrainzId: String?
    
    @Field(key: "mb_artistid")
    var musicBrainzArtistId: String?
    
    @Field(key: "encoder")
    var encoder: String?
    
    @Field(key: "discogs_artistid")
    var discogsArtistId: Int?
    
    @Field(key: "added")
    var addedAt: Date
    
    @Children(for: \.$item)
    var attributes: [ItemAttribute]
    
    var path: String? {
        String(data: pathData, encoding: .utf8)
    }
    
    
    
    // MARK: - Functions
    
    // MARK: Comparable Functions
    
    static func < (lhs: Item, rhs: Item) -> Bool {
        lhs.orderIndex < rhs.orderIndex
    }
    
    
    // MARK: Equatable Functions
    
    static func == (lhs: Item, rhs: Item) -> Bool {
        lhs.id == rhs.id
    }
}

File: FinchServer/Beets/Item.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Text;

namespace FinchServer.Beets;

[Table("items")]
public class Item {
    
    // - Properties
    
    [Key]
    public int Id { get; set; }
    
    [Column("track")]
    public int? Track { get; set; }
    
    [Column("disc")]
    public int? Disc { get; set; }
    
    [Column("title")]
    public required string Title { get; set; }
    
    [Column("artist")]
    public required string Artist { get; set; }
    
    [Column("path")]
    public required byte[] PathData { get; set; }
    
    [Column("length")]
    public double Length { get; set; }
    
    [Column("format")]
    public required string Format { get; set; }
    
    [Column("bitrate")]
    public int Bitrate { get; set; }
    
    [Column("samplerate")]
    public int SampleRate { get; set; }
    
    [Column("album_id")]
    public int? AlbumId { get; set; }
    
    [Column("albumtype")]
    public required string AlbumType { get; set; }
    
    [Column("genre")]
    public string? Genre { get; set; }
    
    [Column("style")]
    public string? Style { get; set; }
    
    [Column("added")]
    public required double Added { get; set; }
    
    [Column("discogs_artistid")]
    public int? DiscogsArtistId { get; set; }
    
    public Album? Album { get; set; }
    
    public string Path => Encoding.UTF8.GetString(PathData);
    public TimeSpan Duration => TimeSpan.FromSeconds(Length);

    public string DurationDescription {
        get {
            var duration = Duration;
            return duration.Hours > 0
                ? $"{duration.Hours:D2}:{duration.Minutes:D2}:{duration.Seconds:D2}"
                : $"{duration.Minutes:D2}:{duration.Seconds:D2}";
        }
    }
}